huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go

```diff
@@ -330,9 +334,182 @@ func resourceIpAddressGroupUpdate(ctx context.Context, d *schema.ResourceData, m
 		}
 	}
 
+	if d.HasChange("ip_addresses") {
+		oldIpsRaw, newIpsRaw := d.GetChange("ip_addresses")
+		addIpsRaw := newIpsRaw.(*schema.Set).Difference(oldIpsRaw.(*schema.Set))
+		removeIpsRaw := oldIpsRaw.(*schema.Set).Difference(newIpsRaw.(*schema.Set))
+
+		if removeIpsRaw.Len() > 0 {
+			err = removeIps(ctx, d, meta, updateIpAddressGroupClient, removeIpsRaw.List())
+			if err != nil {
+				return diag.FromErr(err)
+			}
+		}
+
+		if addIpsRaw.Len() > 0 {
+			err = addIps(ctx, d, meta, updateIpAddressGroupClient, addIpsRaw.List())
+			if err != nil {
+				return diag.FromErr(err)
+			}
+		}
+	}
+
+	if d.HasChange("listeners") {
+		oldListenerRaw, newListenerRaw := d.GetChange("listeners")
+		associateListenerRaw := newListenerRaw.(*schema.Set).Difference(oldListenerRaw.(*schema.Set))
+		disassociateListenerRaw := oldListenerRaw.(*schema.Set).Difference(newListenerRaw.(*schema.Set))
+
+		if disassociateListenerRaw.Len() > 0 {
+			err = disassociateListener(ctx, d, meta, updateIpAddressGroupClient, disassociateListenerRaw.List())
+			if err != nil {
+				return diag.FromErr(err)
+			}
+		}
+
+		if associateListenerRaw.Len() > 0 {
+			err = associateListener(ctx, d, meta, updateIpAddressGroupClient, associateListenerRaw.List())
+			if err != nil {
+				return diag.FromErr(err)
+			}
+		}
+	}
+
 	return resourceIpAddressGroupRead(ctx, d, meta)
 }
 
+func addIps(ctx context.Context, d *schema.ResourceData, meta interface{}, client *golangsdk.ServiceClient,
+	rawParams interface{}) error {
+	addIpsHttpUrl := "v1/ip-groups/{ip_group_id}/add-ips"
+	addIpsPath := client.Endpoint + addIpsHttpUrl
+	addIpsPath = strings.ReplaceAll(addIpsPath, "{ip_group_id}", d.Id())
+
+	addIpsOpt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+	}
+
+	if rawArray, ok := rawParams.([]interface{}); ok {
+		rst := make([]map[string]interface{}, len(rawArray))
+		for i, v := range rawArray {
+			raw := v.(map[string]interface{})
+			rst[i] = map[string]interface{}{
+				"cidr":        utils.ValueIngoreEmpty(raw["cidr"]),
```

Comment:
```suggestion
				"cidr":        raw["cidr"],
```

Reply:
this field is requied, don't need ingore empty

Reply:
done